File: src/main/java/io/schell/llm/benchmark/agent/PiMessageProcessor.java
package io.schell.llm.benchmark.agent;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.function.Consumer;

/**
 * Message processor for Pi agent JSON event stream output.
 * Parses and formats Pi events for console display, similar to ClaudeMessageProcessor.
 */
public class PiMessageProcessor implements Consumer<String> {
    private final ObjectMapper objectMapper = new ObjectMapper();
    private final Consumer<String> outputConsumer;

    public PiMessageProcessor(Consumer<String> outputConsumer) {
        this.outputConsumer = outputConsumer;
    }

    @Override
    public void accept(String s) {
        try {
            JsonNode jsonNode = objectMapper.readTree(s);
            String type = jsonNode.has("type") ? jsonNode.get("type").asText() : null;

            if (type == null) {
                println(s);
                return;
            }

            switch (type) {
                case "session" -> {
                    // Session header - show session info
                    String id = jsonNode.has("id") ? jsonNode.get("id").asText() : "unknown";
                    String cwd = jsonNode.has("cwd") ? jsonNode.get("cwd").asText() : ".";
                    int version = jsonNode.has("version") ? jsonNode.get("version").asInt(1) : 1;
                    println(String.format("[Session v%d] ID: %s, CWD: %s", version, id, cwd));
                }

                case "agent_start" -> {
                    println("\n[Agent starting...]");
                }

                case "agent_end" -> {
                    println("\n[Agent finished]");
                }

                case "turn_start" -> {
                    println("\n--- Turn Start ---");
                }

                case "turn_end" -> {
                    println("--- Turn End ---\n");
                }

                case "message_start" -> {
                    handleMessageStart(jsonNode);
                }

                case "message_update" -> {
                    handleMessageUpdate(jsonNode);
                }

                case "message_end" -> {
                    // Message complete, add newline
                    println("");
                }

                case "tool_execution_start" -> {
                    handleToolExecutionStart(jsonNode);
                }

                case "tool_execution_update" -> {
                    handleToolExecutionUpdate(jsonNode);
                }

                case "tool_execution_end" -> {
                    handleToolExecutionEnd(jsonNode);
                }

                case "auto_compaction_start" -> {
                    String reason = jsonNode.has("reason") ? jsonNode.get("reason").asText() : "unknown";
                    println(String.format("\n[Auto-compaction started: %s]", reason));
                }

                case "auto_compaction_end" -> {
                    boolean aborted = jsonNode.has("aborted") && jsonNode.get("aborted").asBoolean();
                    boolean willRetry = jsonNode.has("willRetry") && jsonNode.get("willRetry").asBoolean();
                    if (aborted) {
                        println("[Auto-compaction aborted]");
                    } else if (willRetry) {
                        println("[Auto-compaction failed, will retry]");
                    } else {
                        println("[Auto-compaction completed]");
                    }
                }

                case "auto_retry_start" -> {
                    int attempt = jsonNode.has("attempt") ? jsonNode.get("attempt").asInt() : 1;
                    int maxAttempts = jsonNode.has("maxAttempts") ? jsonNode.get("maxAttempts").asInt() : 3;
                    String errorMessage = jsonNode.has("errorMessage") ? jsonNode.get("errorMessage").asText() : "";
                    println(String.format("\n[Auto-retry %d/%d: %s]", attempt, maxAttempts, errorMessage));
                }

                case "auto_retry_end" -> {
                    boolean success = jsonNode.has("success") && jsonNode.get("success").asBoolean();
                    int attempt = jsonNode.has("attempt") ? jsonNode.get("attempt").asInt() : 1;
                    if (success) {
                        println(String.format("[Auto-retry %d succeeded]", attempt));
                    } else {
                        String finalError = jsonNode.has("finalError") ? jsonNode.get("finalError").asText() : "unknown error";
                        println(String.format("[Auto-retry %d failed: %s]", attempt, finalError));
                    }
                }

                case "model_change" -> {
                    String provider = jsonNode.has("provider") ? jsonNode.get("provider").asText() : "unknown";
                    String modelId = jsonNode.has("modelId") ? jsonNode.get("modelId").asText() : "unknown";
                    println(String.format("\n[Model changed to: %s/%s]", provider, modelId));
                }

                case "thinking_level_change" -> {
                    String thinkingLevel = jsonNode.has("thinkingLevel") ? jsonNode.get("thinkingLevel").asText() : "unknown";
                    println(String.format("\n[Thinking level: %s]", thinkingLevel));
                }

                case "compaction" -> {
                    String summary = jsonNode.has("summary") ? jsonNode.get("summary").asText() : "";
                    int tokensBefore = jsonNode.has("tokensBefore") ? jsonNode.get("tokensBefore").asInt() : 0;
                    println(String.format("\n[Compacted %d tokens: %s]", tokensBefore, summary.substring(0, Math.min(100, summary.length()))));
                }

                case "branch_summary" -> {
                    String summary = jsonNode.has("summary") ? jsonNode.get("summary").asText() : "";
                    println(String.format("\n[Branched: %s]", summary.substring(0, Math.min(100, summary.length()))));
                }

                case "custom_message" -> {
                    String customType = jsonNode.has("customType") ? jsonNode.get("customType").asText() : "extension";
                    String content = jsonNode.has("content") ? jsonNode.get("content").asText() : "";
                    boolean display = jsonNode.has("display") && jsonNode.get("display").asBoolean();
                    if (display) {
                        println(String.format("\n[%s] %s", customType, content));
                    }
                }

                case "bashExecution" -> {
                    handleBashExecution(jsonNode);
                }

                default -> {
                    // Unknown event type, print raw JSON
                    println(s);
                }
            }
        } catch (JsonProcessingException e) {
            // Not valid JSON, print as-is (might be warnings or other output)
            println(s);
        } catch (Exception e) {
            // Handle any parsing errors gracefully
            println("[Error processing event: " + e.getMessage() + "]");
        }
    }

    private void handleMessageStart(JsonNode jsonNode) {
        JsonNode message = jsonNode.has("message") ? jsonNode.get("message") : null;
        if (message == null) return;

        String role = message.has("role") ? message.get("role").asText() : "unknown";
        JsonNode content = message.has("content") ? message.get("content") : null;

        switch (role) {
            case "user" -> {
                if (content != null && content.isTextual()) {
                    println("\n[User]: " + content.asText());
                } else if (content != null && content.isArray()) {
                    printUserContent(content);
                }
            }

            case "assistant" -> {
                println("\n[Assistant]:");
                if (content != null && content.isArray()) {
                    for (JsonNode item : content) {
                        String itemType = item.has("type") ? item.get("type").asText() : "";
                        switch (itemType) {
                            case "text" -> {
                                String text = item.has("text") ? item.get("text").asText() : "";
                                print(text);
                            }
                            case "thinking" -> {
                                String thinking = item.has("thinking") ? item.get("thinking").asText() : "";
                                println("\n[Thinking]:");
                                println(thinking);
                            }
                            case "toolCall" -> {
                                handleToolCall(item);
                            }
                        }
                    }
                } else if (content != null && content.isObject()) {
                    String itemType = content.has("type") ? content.get("type").asText() : "";
                    switch (itemType) {
                        case "text" -> {
                            String text = content.has("text") ? content.get("text").asText() : "";
                            print(text);
                        }
                        case "thinking" -> {
                            String thinking = content.has("thinking") ? content.get("thinking").asText() : "";
                            println("\n[Thinking]:");
                            println(thinking);
                        }
                        case "toolCall" -> {
                            handleToolCall(content);
                        }
                    }
                }
            }

            case "toolResult" -> {
                if (content != null && content.isArray()) {
                    printToolResultContent(content);
                } else if (content != null && content.isTextual()) {
                    println("[Tool Result]: " + content.asText());
                }
            }

            case "bashExecution" -> {
                handleBashExecution(message);
            }
        }
    }

    private void handleMessageUpdate(JsonNode jsonNode) {
        JsonNode assistantMessageEvent = jsonNode.has("assistantMessageEvent") ? jsonNode.get("assistantMessageEvent") : null;
        if (assistantMessageEvent == null) return;

        String eventType = assistantMessageEvent.has("type") ? assistantMessageEvent.get("type").asText() : "";

        switch (eventType) {
            case "text_delta" -> {
                String delta = assistantMessageEvent.has("delta") ? assistantMessageEvent.get("delta").asText() : "";
                print(delta);
            }

            case "thinking_delta" -> {
                String delta = assistantMessageEvent.has("delta") ? assistantMessageEvent.get("delta").asText() : "";
                print(delta);
            }

            case "tool_call_delta" -> {
                // Tool call in progress, skip for now
            }

            default -> {
                // Other update types, ignore or print raw
            }
        }
    }

    private void handleToolExecutionStart(JsonNode jsonNode) {
        String toolName = jsonNode.has("toolName") ? jsonNode.get("toolName").asText() : "unknown";
        String toolCallId = jsonNode.has("toolCallId") ? jsonNode.get("toolCallId").asText() : "";
        JsonNode args = jsonNode.has("args") ? jsonNode.get("args") : null;

        println(String.format("\n[Tool: %s] (ID: %s)", toolName, toolCallId));

        if (args != null && !args.isNull()) {
            printToolArgs(toolName, args);
        }
    }

    private void handleToolExecutionUpdate(JsonNode jsonNode) {
        String toolName = jsonNode.has("toolName") ? jsonNode.get("toolName").asText() : "unknown";
        JsonNode partialResult = jsonNode.has("partialResult") ? jsonNode.get("partialResult") : null;

        if (partialResult != null && !partialResult.isNull()) {
            println(String.format("[Tool %s progress]...", toolName));
        }
    }

    private void handleToolExecutionEnd(JsonNode jsonNode) {
        String toolName = jsonNode.has("toolName") ? jsonNode.get("toolName").asText() : "unknown";
        boolean isError = jsonNode.has("isError") && jsonNode.get("isError").asBoolean();
        JsonNode result = jsonNode.has("result") ? jsonNode.get("result") : null;

        if (isError) {
            println(String.format("[Tool %s failed]", toolName));
        } else {
            println(String.format("[Tool %s completed]", toolName));
        }

        if (result != null && !result.isNull()) {
            printToolResult(toolName, result);
        }
    }

    private void handleBashExecution(JsonNode jsonNode) {
        String command = jsonNode.has("command") ? jsonNode.get("command").asText() : "";
        String output = jsonNode.has("output") ? jsonNode.get("output").asText() : "";
        Integer exitCode = jsonNode.has("exitCode") && !jsonNode.get("exitCode").isNull() 
                ? jsonNode.get("exitCode").asInt() : null;
        boolean cancelled = jsonNode.has("cancelled") && jsonNode.get("cancelled").asBoolean();
        boolean truncated = jsonNode.has("truncated") && jsonNode.get("truncated").asBoolean();

        println(String.format("\n[Bash: %s]", command));

        if (output != null && !output.isEmpty()) {
            // Limit output length for display
            String displayOutput = output.length() > 500 ? output.substring(0, 500) + "...[truncated]" : output;
            println(displayOutput);
        }

        if (exitCode != null) {
            println(String.format("[Exit code: %d]", exitCode));
        }

        if (cancelled) {
            println("[Command cancelled]");
        }

        if (truncated) {
            println("[Output truncated]");
        }
    }

    private void handleToolCall(JsonNode item) {
        String name = item.has("name") ? item.get("name").asText() : "unknown";
        JsonNode arguments = item.has("arguments") ? item.get("arguments") : null;

        println(String.format("[Tool Call: %s]", name));

        if (arguments != null && !arguments.isNull()) {
            printToolArgs(name, arguments);
        }
    }

    private void printToolArgs(String toolName, JsonNode args) {
        switch (toolName) {
            case "Read" -> {
                String filePath = args.has("file_path") ? args.get("file_path").asText() : "";
                println(String.format("  Reading: %s", filePath));
            }

            case "Write" -> {
                String filePath = args.has("file_path") ? args.get("file_path").asText() : "";
                println(String.format("  Writing: %s", filePath));
            }

            case "Edit" -> {
                String filePath = args.has("file_path") ? args.get("file_path").asText() : "";
                println(String.format("  Editing: %s", filePath));
            }

            case "Bash" -> {
                String command = args.has("command") ? args.get("command").asText() : "";
                println(String.format("  Command: %s", command));
            }

            case "Glob" -> {
                String pattern = args.has("pattern") ? args.get("pattern").asText() : "";
                println(String.format("  Pattern: %s", pattern));
            }

            case "Grep" -> {
                String pattern = args.has("pattern") ? args.get("pattern").asText() : "";
                String filePath = args.has("file_path") ? args.get("file_path").asText() : "";
                println(String.format("  Pattern: %s, File: %s", pattern, filePath));
            }

            default -> {
                // Print raw JSON for unknown tools
                println("  Args: " + args.toPrettyString());
            }
        }
    }

    private void printToolResult(String toolName, JsonNode result) {
        if (result.isTextual()) {
            String content = result.asText();
            // Limit output length
            if (content.length() > 1000) {
                println("  Output: " + content.substring(0, 1000) + "...[truncated]");
            } else {
                println("  Output: " + content);
            }
        } else if (result.isObject()) {
            // For complex results, show summary
            println("  Result: " + result.toPrettyString());
        }
    }

    private void printToolResultContent(JsonNode content) {
        for (JsonNode item : content) {
            String itemType = item.has("type") ? item.get("type").asText() : "";
            switch (itemType) {
                case "text" -> {
                    String text = item.has("text") ? item.get("text").asText() : "";
                    println("[Result]: " + text);
                }
                case "image" -> {
                    println("[Result: image]");
                }
            }
        }
    }

    private void printUserContent(JsonNode content) {
        for (JsonNode item : content) {
            String itemType = item.has("type") ? item.get("type").asText() : "";
            switch (itemType) {
                case "text" -> {
                    String text = item.has("text") ? item.get("text").asText() : "";
                    println("[User]: " + text);
                }
                case "image" -> {
                    println("[User: image]");
                }
            }
        }
    }

    private void print(String s) {
        System.out.print(s);
        if (outputConsumer != null) {
            outputConsumer.accept(s);
        }
    }

    private void println(String s) {
        System.out.println(s);
        if (outputConsumer != null) {
            outputConsumer.accept(s + "\n");
        }
    }
}

File: crates/benchmark-core/src/agent/pi_message_processor.rs
//! PiMessageProcessor - parses Pi agent JSON event stream output
//! Logs key events with tracing and forwards formatted output to consumer for web UI streaming.

use serde_json::Value;
use tracing::{info, debug, warn};

/// Callback type for sending processed output to the web UI SSE stream.
pub type OutputConsumer = Box<dyn Fn(&str) + Send + Sync>;

/// Processes Pi agent's JSON event stream output.
/// Parses structured events and logs key information with tracing.
pub struct PiMessageProcessor {
    consumer: Option<OutputConsumer>,
}

impl PiMessageProcessor {
    /// Create a new processor with an optional output consumer.
    pub fn new(consumer: Option<OutputConsumer>) -> Self {
        Self { consumer }
    }

    /// Process a single line of output from Pi agent.
    pub fn process(&self, line: &str) {
        if let Ok(json) = serde_json::from_str::<Value>(line) {
            if let Some(event_type) = json.get("type").and_then(|v| v.as_str()) {
                match event_type {
                    "session" => self.handle_session(&json),
                    "agent_start" => {
                        info!("[Pi] Agent starting...");
                        self.send_output("\n[Agent starting...]\n");
                    }
                    "agent_end" => {
                        info!("[Pi] Agent finished");
                        self.send_output("\n[Agent finished]\n");
                    }
                    "turn_start" => {
                        info!("[Pi] Turn start");
                        self.send_output("\n--- Turn Start ---\n");
                    }
                    "turn_end" => self.send_output("--- Turn End ---\n"),
                    "message_start" => self.handle_message_start(&json),
                    "message_update" => self.handle_message_update(&json),
                    "message_end" => self.send_output("\n"),
                    "tool_execution_start" => self.handle_tool_execution_start(&json),
                    "tool_execution_update" => self.handle_tool_execution_update(&json),
                    "tool_execution_end" => self.handle_tool_execution_end(&json),
                    "auto_compaction_start" => self.handle_auto_compaction_start(&json),
                    "auto_compaction_end" => self.handle_auto_compaction_end(&json),
                    "auto_retry_start" => self.handle_auto_retry_start(&json),
                    "auto_retry_end" => self.handle_auto_retry_end(&json),
                    "model_change" => self.handle_model_change(&json),
                    "thinking_level_change" => self.handle_thinking_level_change(&json),
                    "compaction" => self.handle_compaction(&json),
                    "branch_summary" => self.handle_branch_summary(&json),
                    "custom_message" => self.handle_custom_message(&json),
                    "bashExecution" => self.handle_bash_execution(&json),
                    _ => {
                        debug!("[Pi unknown event] type={}", event_type);
                        self.send_output(line);
                    }
                }
            } else {
                self.send_output(line);
            }
        } else {
            // Not JSON - might be warnings or other output
            info!("[Pi output] {}", line);
            self.send_output(line);
        }
    }

    fn handle_session(&self, json: &Value) {
        let id = json.get("id").and_then(|v| v.as_str()).unwrap_or("unknown");
        let cwd = json.get("cwd").and_then(|v| v.as_str()).unwrap_or(".");
        info!("[Pi Session] ID: {}, CWD: {}", id, cwd);
        self.send_output(&format!("[Session] ID: {}, CWD: {}\n", id, cwd));
    }

    fn handle_message_start(&self, json: &Value) {
        if let Some(message) = json.get("message").and_then(|v| v.as_object()) {
            if let Some(role) = message.get("role").and_then(|v| v.as_str()) {
                match role {
                    "user" => {
                        if let Some(content) = message.get("content") {
                            if content.is_string() {
                                let text = content.as_str().unwrap_or("");
                                info!("[Pi User] {}", &text[..text.len().min(200)]);
                                self.send_output(&format!("\n[User]: {}\n", text));
                            } else if content.is_array() {
                                if let Some(items) = content.as_array() {
                                    for item in items {
                                        if let Some(item_type) = item.get("type").and_then(|v| v.as_str()) {
                                            match item_type {
                                                "text" => {
                                                    if let Some(text) = item.get("text").and_then(|v| v.as_str()) {
                                                        self.send_output(&format!("[User]: {}\n", text));
                                                    }
                                                }
                                                _ => {}
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                    "assistant" => {
                        info!("[Pi Assistant]");
                        self.send_output("\n[Assistant]:\n");
                        if let Some(content) = message.get("content") {
                            if content.is_array() {
                                if let Some(items) = content.as_array() {
                                    for item in items {
                                        if let Some(item_type) = item.get("type").and_then(|v| v.as_str()) {
                                            match item_type {
                                                "text" => {
                                                    if let Some(text) = item.get("text").and_then(|v| v.as_str()) {
                                                        self.send_output(text);
                                                    }
                                                }
                                                "thinking" => {
                                                    if let Some(thinking) = item.get("thinking").and_then(|v| v.as_str()) {
                                                        self.send_output(&format!("\n[Thinking]:\n{}\n", thinking));
                                                    }
                                                }
                                                "toolCall" => self.handle_tool_call(item),
                                                _ => {}
                                            }
                                        }
                                    }
                                }
                            } else if content.is_object() {
                                if let Some(item_type) = content.get("type").and_then(|v| v.as_str()) {
                                    match item_type {
                                        "text" => {
                                            if let Some(text) = content.get("text").and_then(|v| v.as_str()) {
                                                self.send_output(text);
                                            }
                                        }
                                        "thinking" => {
                                            if let Some(thinking) = content.get("thinking").and_then(|v| v.as_str()) {
                                                self.send_output(&format!("\n[Thinking]:\n{}\n", thinking));
                                            }
                                        }
                                        "toolCall" => self.handle_tool_call(content),
                                        _ => {}
                                    }
                                }
                            }
                        }
                    }
                    "toolResult" => {
                        if let Some(content) = message.get("content") {
                            if content.is_string() {
                                let text = content.as_str().unwrap_or("");
                                info!("[Pi Tool Result] {}", &text[..text.len().min(500)]);
                                self.send_output(&format!("[Tool Result]: {}\n", text));
                            } else if content.is_array() {
                                if let Some(items) = content.as_array() {
                                    for item in items {
                                        if let Some(item_type) = item.get("type").and_then(|v| v.as_str()) {
                                            match item_type {
                                                "text" => {
                                                    if let Some(text) = item.get("text").and_then(|v| v.as_str()) {
                                                        self.send_output(&format!("[Result]: {}\n", text));
                                                    }
                                                }
                                                _ => {}
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                    _ => {}
                }
            }
        }
    }

    fn handle_message_update(&self, json: &Value) {
        if let Some(assistant_event) = json.get("assistantMessageEvent").and_then(|v| v.as_object()) {
            if let Some(event_type) = assistant_event.get("type").and_then(|v| v.as_str()) {
                match event_type {
                    "text_delta" => {
                        if let Some(delta) = assistant_event.get("delta").and_then(|v| v.as_str()) {
                            self.send_output(delta);
                        }
                    }
                    "thinking_delta" => {
                        if let Some(delta) = assistant_event.get("delta").and_then(|v| v.as_str()) {
                            self.send_output(delta);
                        }
                    }
                    "tool_call_delta" => {
                        // Tool call in progress - skip for cleaner output
                    }
                    _ => {}
                }
            }
        }
    }

    fn handle_tool_execution_start(&self, json: &Value) {
        let tool_name = json.get("toolName").and_then(|v| v.as_str()).unwrap_or("unknown");
        let tool_call_id = json.get("toolCallId").and_then(|v| v.as_str()).unwrap_or("");
        info!("[Pi Tool: {}] (ID: {})", tool_name, tool_call_id);
        self.send_output(&format!("\n[Tool: {}] (ID: {})\n", tool_name, tool_call_id));

        if let Some(args) = json.get("args") {
            self.print_tool_args(tool_name, args);
        }
    }

    fn handle_tool_execution_update(&self, json: &Value) {
        let tool_name = json.get("toolName").and_then(|v| v.as_str()).unwrap_or("unknown");
        if json.get("partialResult").is_some() {
            info!("[Pi Tool {}] progress...", tool_name);
            self.send_output(&format!("[Tool {} progress]...\n", tool_name));
        }
    }

    fn handle_tool_execution_end(&self, json: &Value) {
        let tool_name = json.get("toolName").and_then(|v| v.as_str()).unwrap_or("unknown");
        let is_error = json.get("isError").and_then(|v| v.as_bool()).unwrap_or(false);
        if is_error {
            warn!("[Pi Tool {}] failed", tool_name);
            self.send_output(&format!("[Tool {} failed]\n", tool_name));
        } else {
            info!("[Pi Tool {}] completed", tool_name);
            self.send_output(&format!("[Tool {} completed]\n", tool_name));
        }
        if let Some(result) = json.get("result") {
            self.print_tool_result(tool_name, result);
        }
    }

    fn handle_tool_call(&self, item: &Value) {
        if let Some(name) = item.get("name").and_then(|v| v.as_str()) {
            info!("[Pi Tool Call: {}]", name);
            self.send_output(&format!("[Tool Call: {}]\n", name));
            if let Some(arguments) = item.get("arguments") {
                self.print_tool_args(name, arguments);
            }
        }
    }

    fn print_tool_args(&self, tool_name: &str, args: &Value) {
        match tool_name {
            "Read" => {
                if let Some(file_path) = args.get("file_path").and_then(|v| v.as_str()) {
                    info!("  Reading: {}", file_path);
                    self.send_output(&format!("  Reading: {}\n", file_path));
                }
            }
            "Write" => {
                if let Some(file_path) = args.get("file_path").and_then(|v| v.as_str()) {
                    info!("  Writing: {}", file_path);
                    self.send_output(&format!("  Writing: {}\n", file_path));
                }
            }
            "Edit" => {
                if let Some(file_path) = args.get("file_path").and_then(|v| v.as_str()) {
                    info!("  Editing: {}", file_path);
                    self.send_output(&format!("  Editing: {}\n", file_path));
                }
            }
            "Bash" => {
                if let Some(command) = args.get("command").and_then(|v| v.as_str()) {
                    info!("  Command: {}", command);
                    self.send_output(&format!("  Command: {}\n", command));
                }
            }
            "Glob" => {
                if let Some(pattern) = args.get("pattern").and_then(|v| v.as_str()) {
                    info!("  Pattern: {}", pattern);
                    self.send_output(&format!("  Pattern: {}\n", pattern));
                }
            }
            "Grep" => {
                if let Some(pattern) = args.get("pattern").and_then(|v| v.as_str()) {
                    if let Some(file_path) = args.get("file_path").and_then(|v| v.as_str()) {
                        info!("  Pattern: {} in {}", pattern, file_path);
                        self.send_output(&format!("  Pattern: {} in {}\n", pattern, file_path));
                    }
                }
            }
            _ => {
                debug!("  Args: {}", args);
                self.send_output(&format!("  Args: {}\n", args));
            }
        }
    }

    fn print_tool_result(&self, _tool_name: &str, result: &Value) {
        if result.is_string() {
            let content = result.as_str().unwrap_or("");
            let output_text = if content.len() > 1000 {
                format!("{}...[truncated]", &content[..1000])
            } else {
                content.to_string()
            };
            info!("  Output: {}", output_text);
            self.send_output(&format!("  Output: {}\n", output_text));
        }
    }

    fn handle_auto_compaction_start(&self, json: &Value) {
        let reason = json.get("reason").and_then(|v| v.as_str()).unwrap_or("unknown");
        info!("[Pi Auto-compaction started: {}]", reason);
        self.send_output(&format!("\n[Auto-compaction started: {}]\n", reason));
    }

    fn handle_auto_compaction_end(&self, json: &Value) {
        let aborted = json.get("aborted").and_then(|v| v.as_bool()).unwrap_or(false);
        let will_retry = json.get("willRetry").and_then(|v| v.as_bool()).unwrap_or(false);
        if aborted {
            info!("[Pi Auto-compaction aborted]");
            self.send_output("[Auto-compaction aborted]\n");
        } else if will_retry {
            info!("[Pi Auto-compaction failed, will retry]");
            self.send_output("[Auto-compaction failed, will retry]\n");
        } else {
            info!("[Pi Auto-compaction completed]");
            self.send_output("[Auto-compaction completed]\n");
        }
    }

    fn handle_auto_retry_start(&self, json: &Value) {
        let attempt = json.get("attempt").and_then(|v| v.as_i64()).unwrap_or(1);
        let max_attempts = json.get("maxAttempts").and_then(|v| v.as_i64()).unwrap_or(3);
        let error_message = json.get("errorMessage").and_then(|v| v.as_str()).unwrap_or("");
        info!("[Pi Auto-retry {}/{}: {}]", attempt, max_attempts, error_message);
        self.send_output(&format!("\n[Auto-retry {}/{}: {}]\n", attempt, max_attempts, error_message));
    }

    fn handle_auto_retry_end(&self, json: &Value) {
        let success = json.get("success").and_then(|v| v.as_bool()).unwrap_or(false);
        let attempt = json.get("attempt").and_then(|v| v.as_i64()).unwrap_or(1);
        if success {
            info!("[Pi Auto-retry {} succeeded]", attempt);
            self.send_output(&format!("[Auto-retry {} succeeded]\n", attempt));
        } else {
            let final_error = json.get("finalError").and_then(|v| v.as_str()).unwrap_or("unknown error");
            warn!("[Pi Auto-retry {} failed: {}]", attempt, final_error);
            self.send_output(&format!("[Auto-retry {} failed: {}]\n", attempt, final_error));
        }
    }

    fn handle_model_change(&self, json: &Value) {
        let provider = json.get("provider").and_then(|v| v.as_str()).unwrap_or("unknown");
        let model_id = json.get("modelId").and_then(|v| v.as_str()).unwrap_or("unknown");
        info!("[Pi Model changed to: {}/{}]", provider, model_id);
        self.send_output(&format!("\n[Model changed to: {}/{}]\n", provider, model_id));
    }

    fn handle_thinking_level_change(&self, json: &Value) {
        let level = json.get("thinkingLevel").and_then(|v| v.as_str()).unwrap_or("unknown");
        info!("[Pi Thinking level: {}]", level);
        self.send_output(&format!("\n[Thinking level: {}]\n", level));
    }

    fn handle_compaction(&self, json: &Value) {
        let summary = json.get("summary").and_then(|v| v.as_str()).unwrap_or("");
        let tokens_before = json.get("tokensBefore").and_then(|v| v.as_i64()).unwrap_or(0);
        let preview = &summary[..summary.len().min(100)];
        info!("[Pi Compacted {} tokens: {}]", tokens_before, preview);
        self.send_output(&format!("\n[Compacted {} tokens: {}]\n", tokens_before, preview));
    }

    fn handle_branch_summary(&self, json: &Value) {
        let summary = json.get("summary").and_then(|v| v.as_str()).unwrap_or("");
        let preview = &summary[..summary.len().min(100)];
        info!("[Pi Branched: {}]", preview);
        self.send_output(&format!("\n[Branched: {}]\n", preview));
    }

    fn handle_custom_message(&self, json: &Value) {
        let custom_type = json.get("customType").and_then(|v| v.as_str()).unwrap_or("extension");
        let content = json.get("content").and_then(|v| v.as_str()).unwrap_or("");
        let display = json.get("display").and_then(|v| v.as_bool()).unwrap_or(false);
        if display {
            info!("[Pi {}] {}", custom_type, content);
            self.send_output(&format!("\n[{}] {}\n", custom_type, content));
        }
    }

    fn handle_bash_execution(&self, json: &Value) {
        let message = json.get("message").or(json.get("data")).unwrap_or(json);
        let command = message.get("command").and_then(|v| v.as_str()).unwrap_or("");
        let output = message.get("output").and_then(|v| v.as_str()).unwrap_or("");
        let exit_code = message.get("exitCode").and_then(|v| v.as_i64());
        let cancelled = message.get("cancelled").and_then(|v| v.as_bool()).unwrap_or(false);
        let truncated = message.get("truncated").and_then(|v| v.as_bool()).unwrap_or(false);

        info!("[Pi Bash: {}]", command);
        self.send_output(&format!("\n[Bash: {}]\n", command));

        if !output.is_empty() {
            let display = if output.len() > 500 {
                format!("{}...[truncated]", &output[..500])
            } else {
                output.to_string()
            };
            self.send_output(&format!("{}\n", display));
        }

        if let Some(code) = exit_code {
            self.send_output(&format!("[Exit code: {}]\n", code));
        }
        if cancelled {
            self.send_output("[Command cancelled]\n");
        }
        if truncated {
            self.send_output("[Output truncated]\n");
        }
    }

    fn send_output(&self, text: &str) {
        if let Some(ref consumer) = self.consumer {
            consumer(text);
        }
    }
}
